apheleia.el

```diff
@@ -447,15 +459,49 @@ modified from what is written to disk, then don't do anything."
                                     output-fname
                                   arg))
                               command)))
+      `(,input-fname ,output-fname ,@command))))
+
+(defun apheleia--run-formatters (commands callback &optional stdin)
+  "Run one or more code formatters on the current buffer.
+The formatter is specified by the COMMANDS list. Each entry in
+COMMANDS should be a list of strings or symbols (see
+`apheleia-format-buffer'). Once all the formatters in COMMANDS
+finish succesfully then invoke CALLBACK with one argument, a
+buffer containing the output of all the formatters.
+
+STDIN is a buffer containing the standard input for the first
+formatter in COMMANDS. This should not be supplied by the caller
+and instead is supplied by this command when invoked recursively.
+The stdout of the previous formatter becomes the stdin of the
+next formatter."
+  ;; WARN: Is there a chance that current-buffer will change when invoked
+  ;; recursively? Say when first called current-buffer is the input file
+  ;; buffer, when called again is it still the original file buffer?
+  ;;
+  ;; To summarise: Does emacs persists current-buffer with lexical scoping?
```

Comment:
Yeah, so in principle the current buffer in the callback should be the stdout buffer associated with the process, regardless of what the current buffer was when the process was spawned.

However, it's probably best to explicitly save and track the current buffer in a lexically scoped variable to avoid any surprises. `current-buffer` is dynamically scoped, regardless of whether lexical binding is enabled or not. (Well, actually `current-buffer` is a function, not a variable, but if you were to analogize it to a variable, it would behave more like a dynamically scoped one than a lexically scoped one.)